aead/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+    + Into<Vec<u8>>
+    + Send
+    + Sized
+    + Sync
+    + for<'bytes> TryFrom<&'bytes [u8], Error = AeadError>
+{
+    /// Retrieve the length of the ciphertext, in bytes
+    fn len(&self) -> usize;
+}
+
+/// An explicit cryptographic nonce, which may be optionally used for
+/// authenticated block ciphers.
+pub trait Nonce:
+    Clone
+    + Debug
+    + Display
+    + Into<Vec<u8>>
+    + Send
+    + Sized
+    + Sync
+    + for<'bytes> TryFrom<&'bytes [u8], Error = AeadError>
+{
+    /// The length of an explicit nonce, in bytes.
+    type Len: ArrayLength<u8> + Unsigned;
+}
+
+/// A standard nonce implementation for AEAD algorithms which do not use
+/// explicit nonces.
+#[cfg(feature = "serde")]
+#[derive(
+    Clone, Copy, Debug, Default, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize,
+)]
+pub struct EmptyNonce;
+#[cfg(not(feature = "serde"))]
+#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct EmptyNonce;
+
+/// Output an empty-array style string for the EmptyNonce display.
+impl Display for EmptyNonce {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "[]")
+    }
+}
+
+/// Size restrictions for the Emtpy Nonce type
+impl Nonce for EmptyNonce {
+    type Len = U0;
+}
+
+/// An empty nonce can be converted into an empty vector
+impl Into<Vec<u8>> for EmptyNonce {
+    fn into(self) -> Vec<u8> {
+        Vec::default()
+    }
+}
+
+/// Load this nonce from a byte slice (myst be zero-length)
+impl<'bytes> TryFrom<&'bytes [u8]> for EmptyNonce {
+    type Error = AeadError;
+
+    fn try_from(src: &[u8]) -> Result<Self, Self::Error> {
+        if src.len() != 0 {
+            Err(AeadError::InvalidNonceSize)
+        } else {
+            Ok(EmptyNonce)
+        }
+    }
+}
+
+/// A trait which can support an RFC5116-style authenticated encryption scheme.
+///
+/// In order to reduce the opportunity for misuse, the nonce used for
+/// encryption is meant to be managed internally by the implementation itself.
+pub trait AuthenticatedBlockCipher: Sized + Send + Sync
+where
+    for<'ciphertext> Vec<u8>: From<&'ciphertext Self::Ciphertext>,
+{
+    /// The message encapsulation type
+    type Ciphertext: Ciphertext;
+
+    /// The cryptographic nonce object used for this cipher
+    type Nonce: Nonce;
+
+    /// Encrypts the given plaintext into a new ciphertext object and the nonce
+    fn encrypt<'ad, RngType: CryptoRng + RngCore>(
+        &mut self,
+        csprng: &mut RngType,
+        additional_data: impl Iterator<Item = &'ad [u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Self::Ciphertext, Self::Nonce), AeadError>;
+
+    /// Authenticates the ciphertext, nonce, and additional data, then
+    /// decrypts the ciphertext contents into plaintext.   
+    fn decrypt<'ad>(
+        &self,
+        additional_data: impl Iterator<Item = &'ad [u8]>,
+        nonce: Self::Nonce,
+        ciphertext: Self::Ciphertext,
+    ) -> Result<Vec<u8>, AeadError>;
```

Comment:
As mentioned earlier, I think it'd be nice for the core trait API that algorithm providers `impl` be no-std friendly. The rest of the RustCrypto crates are, so forcing allocation here is "missed it by that much"